### PR TITLE
Adds floating block comments to the outlining spans response

### LINKED ISSUE #22732: getOutliningSpans does not return multiline comment
**TypeScript Version:**  2.8.0-dev.20180320

From https://github.com/Microsoft/vscode/issues/45975

**Search Terms:** 
- getOutliningSpans
- Folding

**Code**
For the js:

```js
/* 
comment
*/
```

Invoke `getOutliningSpans` on TS Server

**Expected behavior:**
Returns span that includes the multline comment

**Actual behavior:**
No outlining spans returned for the document

### PR DESCRIPTION
**Pull request by orta**
Fixes #22732 

When you have a sourcefile which is just a comment, then there are no statement nodes which we can iterate over to grab the comments from. This PR adds an extra check for top level unattached comments and adds an outlining span for them.

It's careful to only handle multiline comments to not clash with the region support.

### CODE REVIEW

File: src/services/outliningElementsCollector.ts [hunk 1]
```diff
@@ -106,6 +107,23 @@ namespace ts.OutliningElementsCollector {
         return regionDelimiterRegExp.exec(lineText);
     }
 
+    function addTopLevelUnattachedCommentSpans(sourceFile: SourceFile, out: Push<OutliningSpan>): void {
+        // Comments which are attached to statements would be included in addNodeOutliningSpans
+        if (sourceFile.statements.length > 0) return;
+
+        // This will instead set up spans for the missing ones
+        forEach(getLeadingCommentRangesOfNode(sourceFile, sourceFile), (range) => {
```
Comment (sandersn) [inline src/services/outliningElementsCollector.ts]: I don't know this code at all, but it seems like this must be similar to the code in addNodeOutliningSpans. Maybe enough to share?

Edit:
*Narrator voice* It wasn't.

**Comment by orta** [comment]
Great points, I've refactored it to be closer to @andrewbranch's idea. That mostly worked, yep.

Turned into a bit of a show because some tests then had filtered and unfiltered outline spans by types, so I needed to make sure that they were all fine at the fourslash test layer.

**Comment by RyanCavanaugh** [comment]
@typescript-bot cherry-pick this to release-3.8

**Comment by RyanCavanaugh** [comment]
@orta can you manually port to 3.8 please? The cherry-pick build failed due to octokit being out of date. Alternatively you can push a master merge to the feature branch and request the cherry-pick again

**Pull request merged**